### PR TITLE
Fix：RunAllTasks should be implemented by passing the number of tasks

### PR DESCRIPTION
**Pull request by dreamlike-ocean**
Motivation:

RunAllTasks should be implemented by passing the number of tasks to run, rather than using a timeout

Modification:

Add an override method to handle running all tasks correctly

Result:

https://github.com/netty/netty/issues/14165

### CODE REVIEW

**Comment by franz1981** [comment]
> RunAllTasks should be implemented by passing the number of tasks to run, rather than using a timeout

duration of such tasks is not known, some could be long, other shorts; why you want to use the number of tasks?

**Comment by dreamlike-ocean** [comment]
> > RunAllTasks should be implemented by passing the number of tasks to run, rather than using a timeout
> 
> duration of such tasks is not known, some could be long, other shorts; why you want to use the number of tasks?

I have checked the code of SingleThreadIoEventLoop::run. 
If the code uses a timeout, the default value might not be enough. Therefore, I believe the intention of the code is to specify the number of tasks to run.

Do you have any better ideas

``` java
    @Override
    protected void run() {
        assert inEventLoop();
        do {
            runIo();
            if (isShuttingDown()) {
                ioHandler.prepareToDestroy();
            }
            runAllTasks(maxTasksPerRun);

            // We should continue with our loop until we either confirmed a shutdown or we can suspend it.
        } while (!confirmShutdown() && !canSuspend());
    }
```

```java

protected static final int DEFAULT_MAX_TASKS_PER_RUN = Math.max(1,
            SystemPropertyUtil.getInt("io.netty.eventLoop.maxTaskPerRun", 1024 * 4))

```

File: common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java [hunk 1]
```diff
@@ -544,7 +544,39 @@ protected boolean runAllTasks(long timeoutNanos) {
...
+
+        long runTasks = 0;
+        long lastExecutionTime;
+        for (;;) {
+            safeExecute(task);
+
+            runTasks ++;
+
+            task = pollTask();
+            if (task == null || runTasks >= maxTasksPerRun) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java:567]:
```suggestion
            if (task == null || runTasks == maxTasksPerRun) {
```

**Comment by normanmaurer** [comment]
@dreamlike-ocean I think you are correct here.

File: common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java [hunk 2]
```diff
@@ -544,7 +544,39 @@ protected boolean runAllTasks(long timeoutNanos) {
     }
 
     /**
-     * Invoked before returning from {@link #runAllTasks()} and {@link #runAllTasks(long)}.
+     * Poll all tasks from the task queue and run them via {@link Runnable#run()} method.
+     *
+     * @return {@code true} if at least one task was run.
+     */
+    protected boolean runAllTasks(int maxTasksPerRun) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java:551]: Let's rename it to something like this and call it from `SingleThreadIoEventLoop`.
```suggestion
    protected boolean runAllTasksWithLimit(int maxTasksPerRun) {
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by normanmaurer** [comment]
Or wait... I think we should better fix it by using a timeout like @franz1981 suggested. Will do a PR shortly

**Comment by normanmaurer** [comment]
Closed in favor of https://github.com/netty/netty/pull/14557

**Pull request closed without merging**